### PR TITLE
538 - Array access after deallocation.

### PR DESCRIPTION
**Pull request by arshad01**
Moved the free(buf) after all accesses to the buf array.

### CODE REVIEW

File: src/syscheckd/run_realtime.c [hunk 1]
```diff
@@ -91,10 +90,15 @@ int realtime_checksumfile(const char *file_name)
...
             send_syscheck_msg(alert_msg);
 
+            free(buf);
+
             return (1);
         } else {
             mdebug2("Discarding '%s': checksum already reported.", file_name);
         }
+
+        free(buf);
```
Comment (vikman90) [inline src/syscheckd/run_realtime.c]: This `free()` would deallocate memory already present in the hash table. The string should only be freed when the hash table is updated (`OSHash_Update()`) like in the first case.

**Comment by vikman90** [comment]
Hi @arshad01,

thank you for this PR. This bug has escaped us because recently freed memory is yet accessible on platforms with the latest stable Glibc, like Ubuntu. In other systems like CentOS 7, the memory may be corrupted fortuitously.

However, I think the second `free()` operation is not correct because it would deallocate memory referenced by the hash table.

Please take a look at this and tell us if you agree.

Thank you for your valuable contribution!

**Comment by vikman90** [comment]
@arshad01 Thank you for updating the PR. That's a great fix!

**Comment by arshad01** [comment]
@vikman90 Thanks for accepting and merging the PR.

**Pull request merged**